Remove user creation and modernize

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This pull request removes user creation from the configuration and modernizes the rest of the configuration.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

User creation should now be managed by [cisagov/cyhy-users-non-admin]. I modernized the configuration while I was here just to make it more consistent with how we do similar configurations in COOL projects.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Automated tests pass.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.


## ✅ Pre-merge checklist ##

- [x] Migrate users to [cisagov/cyhy-users-non-admin] and remove them from the state.
- [x] Destroy this configuration as it is deployed.
- [x] Do a fresh apply to lay down the policies using the new configuration.

[cisagov/cyhy-users-non-admin]: https://github.com/cisagov/cyhy-users-non-admin

mcdonnnj committed:
Remove user creation from the configuration

User creation should instead be handled by
cisagov/cyhy-users-non-admin and once the users are created they can be
added to the group defined in this configuration.

mcdonnnj committed:
Move group membership handling into its own file

Also switch to using for_each instead of count.

mcdonnnj committed:
GNU nano 5.6.1           /home/mcdonnnj/workspace/cisa_repos/bod-18-01-cloudwatch-users/.git/COMMIT_EDITMSG *
Consolidate IAM policies and move to its own file

Consolidate the two policy documents into a single, comprehensive
document and move that policy document into its own file. Update to a
single group policy and adjust it accordingly. Move the log group data
resource into locals.tf while adjusting things.

mcdonnnj committed:
Use IAM policy attachment instead of a group policy

Move the IAM policy creation and switch to an IAM policy instead of an
IAM group policy. This policy is then attached to the group instead.
Also add variables to control the name and description for this IAM
policy.

mcdonnnj committed:
Add a variable to control the IAM group name

mcdonnnj committed:
Use workspace dependent names for unique resources

Append `-production` if in a production workspace or the workspace name
otherwise to the group and policy names in this configuration.

mcdonnnj committed:
Consolidate lambda information variables

Consolidate the lambda_function_names and scan_types variable
information to be sourced from a single bod_lambdas variable.

jsf9k reviewed (approved):
This looks good to me.  Thanks for revamping this one!

dav3r reviewed (approved):
Excellent, looks solid! ✅

mcdonnnj committed:
Get a list of map values before splatting

Splat expressions only work on lists so we need to get a list of the
aws_cloudwatch_log_group objects before we can splat to get all of the
resource ARNs.

mcdonnnj commented:
Note: The original state for this project was in the `default` workspace. I created an appropriate workspace and re-applied in that workspace.

jsf9k commented on bodlambdalogreadaccess_policy.tf:
```diff
@@ -0,0 +1,37 @@
+      "logs:GetLogEvents",
+    ]
+
+    resources = values(data.aws_cloudwatch_log_group.bod_lambda_logs)[*].arn
```
I personally think a list comprehension is clearer than the splat syntax:
```suggestion
    resources = [for k,v in data.aws_cloudwatch_log_group.bod_lambda_logs : v.arn]
```

dav3r replied:
> I personally think a list comprehension is clearer than the splat syntax

Seconded.  We have been gradually de-splatting our Terraform code whenever possible.

mcdonnnj replied:
Works for me. I was on the fence between the two and only went with the splat because it was closer to what was already approved. Please see [531bb24](https://github.com/cisagov/bod-18-01-cloudwatch-users/pull/9/commits/531bb248d40269ee760ab0de2900f6629ee6d5a1) for the change.

jsf9k reviewed (approved):
I made one suggestion, but I'm fine whether you take it or leave it.

dav3r reviewed (approved):
Approved, though I fully support @jsf9k's suggestion.

mcdonnnj committed:
Use a for expression instead of a splat expression

The team consensus is that a splat expression is less clear than a for
expression in a list comprehension.

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

jsf9k reviewed (approved):
_Intensity of approval intensifies_
